gluonnlp/data/stream.py

```diff
@@ -0,0 +1,313 @@
+
+    Parameters
+    ----------
+    file_pattern: str
+        Path to the input text files.
+    encoding : str, default 'utf8'
+        File encoding format.
+    flatten : bool, default False
+        Whether to return all samples as flattened tokens. If True, each sample is a token.
+    skip_empty : bool, default True
+        Whether to skip the empty samples produced from sample_splitters. If False, `bos` and `eos`
+        will be added in empty samples.
+    sample_splitter : function, default str.splitlines
+        A function that splits the dataset string into samples.
+    tokenizer : function or None, default str.split
+        A function that splits each sample string into list of tokens. If None, raw samples are
+        returned according to `sample_splitter`.
+    bos : str or None, default None
+        The token to add at the begining of each sequence. If None, or if tokenizer is not
+        specified, then nothing is added.
+    eos : str or None, default None
+        The token to add at the end of each sequence. If None, or if tokenizer is not
+        specified, then nothing is added.
+    sampler : str, {'sequential', 'random'}, defaults to 'random'
+        The sampler used to sample texts within a file.
+
+        - 'sequential': SequentialSampler
+        - 'random': RandomSampler
+    file_sampler : str, {'sequential', 'random'}, defaults to 'random'
+        The sampler used to sample a file.
+
+        - 'sequential': SequentialSampler
+        - 'random': RandomSampler
+    """
+    def __init__(self, file_pattern, encoding='utf8', flatten=False, skip_empty=True,
+                 sample_splitter=lambda s: s.splitlines(), tokenizer=lambda s: s.split(),
+                 bos=None, eos=None, sampler='random', file_sampler='random'):
+        assert sample_splitter, 'sample_splitter must be specified.'
+        if not isinstance(file_pattern, str):
+            raise TypeError('file_pattern must be str, but got %s'%type(file_pattern))
+        self._file_pattern = os.path.expanduser(file_pattern)
+        self._encoding = encoding
+        self._flatten = flatten
+        self._skip_empty = skip_empty
+        self._sample_splitter = sample_splitter
+        self._tokenizer = tokenizer
+        self._bos = bos
+        self._eos = eos
+        self._sampler = sampler
+        self._file_sampler = file_sampler
+
+    def _get_sampler(self, sampler):
+        assert isinstance(sampler, str), 'Expected sampler to be a str, but got %s'%type(sampler)
+        if sampler == 'random':
+            return RandomSampler
+        if sampler == 'sequential':
+            return SequentialSampler
+        raise ValueError('sampler must be either "random" or "sequential", but got %s'%(sampler))
+
+    def __iter__(self):
+        sampler = self._get_sampler(self._sampler)
+        file_sampler = self._get_sampler(self._file_sampler)
+        # generate file samples
+        files = glob.glob(self._file_pattern)
```

Comment:
Should we use `sorted(glob.glob(self._file_pattern))`?

Reply:
good catch. will update